kafka-connect-transforms/src/main/java/io/tabular/iceberg/connect/transforms/JsonToMapTransform.java

```diff
@@ -0,0 +1,168 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package io.tabular.iceberg.connect.transforms;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.ObjectReader;
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import java.util.Map;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.kafka.common.config.ConfigDef;
+import org.apache.kafka.connect.connector.ConnectRecord;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Struct;
+import org.apache.kafka.connect.sink.SinkRecord;
+import org.apache.kafka.connect.transforms.Transformation;
+import org.apache.kafka.connect.transforms.util.SimpleConfig;
+
+public class JsonToMapTransform<R extends ConnectRecord<R>> implements Transformation<R> {
+
+  public static final String JSON_LEVEL = "transforms.json.root";
+
+  private static final ObjectReader mapper = new ObjectMapper().reader();
+
+  private boolean startAtRoot = false;
+
+  public static final ConfigDef CONFIG_DEF =
+      new ConfigDef()
+          .define(
+              JSON_LEVEL,
+              ConfigDef.Type.BOOLEAN,
+              false,
+              ConfigDef.Importance.MEDIUM,
+              "Boolean value to start at root. False is one level in from the root");
+
+  private static final String ALL_JSON_SCHEMA_FIELD = "payload";
+  private static final Schema JSON_MAP_SCHEMA =
+      SchemaBuilder.struct()
+          .field(
+              ALL_JSON_SCHEMA_FIELD,
+              SchemaBuilder.map(Schema.STRING_SCHEMA, Schema.STRING_SCHEMA).optional().build())
+          .build();
+
+  @Override
+  public R apply(R record) {
+    if (record.value() == null) {
+      return record;
+    } else {
+      return process(record);
+    }
+  }
+
+  private R process(R record) {
+    if (!(record.value() instanceof String)) {
+      // filter out non-string messages
+      return newNullRecord(record);
+    }
+
+    String json = (String) record.value();
```

Comment:
If we had a proper `org.apache.kafka.connect.Schema` we wouldn't be here in the first place given this discussion: https://github.com/apache/iceberg/pull/9641#discussion_r1515260098 and this https://github.com/apache/kafka/pull/5001/ 

Unless they are providing a JsonSchema I am SOL if they are just providing json.

Reply:
Looking at https://github.com/a0x8o/kafka/blob/master/connect/json/src/main/java/org/apache/kafka/connect/json/JsonConverter.java

doesn't look like it does what we want --infer a valueSchema unless it's backed by some schema (avro, jsonschema, etc)

Reply:
> If we had a proper org.apache.kafka.connect.Schema we wouldn't be here in the first place

XD You're right, it doesn't seem like it returns a `org.apache.kafka.connect.Schema` in the `schemas.enable": "false"` case.

Another idea then; the `JsonConverter` doesn't return a `org.apache.kafka.connect.Schema` but it does return valid `org.apache.kafka.connect.data` types. 
Most likely it will return a `java.util.Map<K, V>`.
Your SMT could look at the runtime types of each key and value of that java.util.Map and based on that build only a `org.apache.kafka.connect.Schema`, making sure to specify `org.apache.kafka.connect.Schema.Type.Struct`/`org.apache.kafka.connect.Schema.Type.Map` as appropriate. 
Then it just needs to return a `SinkRecord` that just includes that `java.util.Map<K, V>` and the `org.apache.kafka.connect.Schema` you just constructed. 

I think you might be looking to write a function that looks very similar to this: 
https://github.com/tabular-io/iceberg-kafka-connect/blob/ae729732c4b3597b3773e0d343a29c8303f16c74/kafka-connect/src/main/java/io/tabular/iceberg/connect/data/SchemaUtils.java#L296

Except instead of returning an Iceberg `Type`, you're returning an `org.apache.kafka.connect.Schema.Type`. 
And the only logical difference would that in the case where you see a  `value instanceof Map`, you would return a MapType rather than a StructType. 

---

Note if you have bonkers data e.g. a List consisting of ints and string i.e. `java.util.List<Object>` you would have to  specify in the Schema that it should be a `Array<String>` and the iceberg-kafka-connect-sink code will try to convert all elements to string automatically here: https://github.com/tabular-io/iceberg-kafka-connect/blob/64b824adfa7cc76716b272e3554128d704553f00/kafka-connect/src/main/java/io/tabular/iceberg/connect/data/RecordConverter.java#L368
So you _shouldn't_ need to do any work on the values.

Reply:
Going to poke around in JsonConverter and see what I get back.

Reply:
Poking around:  

it should convert things properly but still needs some cleanup in the following SMT. 

It will create empty keys from empty objects, and array types will be mixed which will have to get cleaned up in the following SMT

![image](https://github.com/tabular-io/iceberg-kafka-connect/assets/160646114/07b1ded9-3cf9-4986-81cb-eeeb43275674)

Reply:
Other problems with their converter: 

- it overflows on Long instead of going to BigInteger -> String 
- it truncates doubles (doesn't go to BigDecimal) (my code also has this problem) 

It has hooks for doing BigDecimal properly but they don't seem to be used.

Reply:
> I don't know if it is actually an improvement over parsing the json nodes ourselves.

I'll leave the decision to you; I'm not as deep into this as you are. 

--- 
> It will create empty keys from empty objects

I'm guessing this is key2 in your example. 
Doesn't look empty to me in the result, looks like an empty Map. 
That seems correct to me, we should put an empty map into iceberg IMO instead of omitting the key entirely.

---
> array types will be mixed which will have to get cleaned up in the following SMT

if array element types are mixed up, we only have one choice; convert all the elements to strings and write to iceberg an Array<String>. 
You shouldn't _need_ to clean anything up in this SMT, you only need to specify in your Schema that this is a `kafka.connect.data.types.Array<String>`. The connector _should_ do the rest of the work of converting each of the mixed types inside the array to string [here](https://github.com/tabular-io/iceberg-kafka-connect/blob/64b824adfa7cc76716b272e3554128d704553f00/kafka-connect/src/main/java/io/tabular/iceberg/connect/data/RecordConverter.java#L368).

---
> it overflows on Long instead of going to BigInteger.
> large numbers to overflow and become -38729732987

Not happy about either of those but if that's what the JsonConverter does, then that's what it does. 
Anyone having an issue with that can take it up with the JsonConverter maintainers IMO if we go this route.

Reply:
I'm going to park the version that requires JsonConverter to be used.  It's a lot of work and awkward.  This version works and gets us to the same place.